File: c#/278-First-Bad-Version.cs
/* The isBadVersion API is defined in the parent class VersionControl.
      bool IsBadVersion(int version); */

// Time: O(log(n))
// Space: O(n)
public class Solution : VersionControl
{
    public int FirstBadVersion(int n)
    {
        int l = 0,
            r = n;
        while (l < r)
        {
            int mid = l + (r - l) / 2;
            if (IsBadVersion(mid))
            {
                r = mid;
            }
            else
            {
                l = mid + 1;
            }
        }
        return l;
    }
}

File: python/278-First-Bad-Version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:


class Solution:
    # Space O(1)
    # Time O(logn)
    def firstBadVersion(self, n: int) -> int:
        left, right = 1, n

        while left < right:
            mid = left + (right - left) // 2
            if isBadVersion(mid):
                right = mid
            else:
                left = mid + 1
        return left
